src/components/MultiHopTrade/hooks/useReceiveAddress.tsx

```diff
@@ -32,65 +33,60 @@ export const useReceiveAddress = ({
-       */
-      if (buyAssetChainId !== buyAssetAccountChainId) {
-        return
-      }
+  // This flag is used to skip the query below and treat missing input as `isLoading` to prevent UI
+  // flashing during state changes. Any of the conditions below returning true should be treated as
+  // "we're not yet ready to determine if a wallet receive address is available"
+  const isInitializing = useMemo(() => {
+    if (!buyAsset || !wallet || !buyAccountId || !buyAccountMetadata) {
+      return true
+    }
 
-      const fetchUnchainedAddress = Boolean(wallet && isLedger(wallet))
+    const buyAssetChainId = buyAsset.chainId
+    const buyAssetAccountChainId = fromAccountId(buyAccountId).chainId
 
-      const receiveAddress = await getReceiveAddress({
-        asset: buyAsset,
-        wallet,
-        accountMetadata: buyAccountMetadata,
-        deviceId: await wallet.getDeviceID(),
-        pubKey: fetchUnchainedAddress ? fromAccountId(buyAccountId).account : undefined,
-      })
-      return receiveAddress
-    },
-    [buyAccountId, buyAccountMetadata, wallet],
-  )
+    /**
+     * do NOT remove
+     * super dangerous - don't use the wrong bip44 params to generate receive addresses
+     */
+    if (buyAssetChainId !== buyAssetAccountChainId) {
+      return true
+    }
+
+    return false
+  }, [buyAccountId, buyAccountMetadata, buyAsset, wallet])
+
+  const { data: walletReceiveAddress, isLoading } = useQuery({
+    queryKey: ['receiveAddress', buyAsset?.assetId, deviceId, buyAccountId],
+    queryFn: isInitializing
+      ? skipToken
+      : async () => {
+          // Already covered in shouldSkip, but TypeScript lyfe mang.
+          if (!buyAsset || !wallet || !buyAccountId || !buyAccountMetadata) {
+            return
+          }
+
+          if (isUtxoAccountId(buyAccountId) && !buyAccountMetadata?.accountType)
+            throw new Error(`Missing accountType for UTXO account ${buyAccountId}`)
+
+          const fetchUnchainedAddress = Boolean(wallet && isLedger(wallet))
+          const walletReceiveAddress = await getReceiveAddress({
+            asset: buyAsset,
+            wallet,
+            accountMetadata: buyAccountMetadata,
+            deviceId,
+            pubKey: fetchUnchainedAddress ? fromAccountId(buyAccountId).account : undefined,
+          })
 
-  // Set the receiveAddress when the buy asset changes
-  // TODO: This belongs in a react query to avoid race conditions.
-  useEffect(() => {
-    if (!buyAsset) return
-    ;(async () => {
-      try {
-        const updatedReceiveAddress = await getReceiveAddressFromBuyAsset(buyAsset)
-        setWalletReceiveAddress(updatedReceiveAddress)
-      } catch (e) {
-        console.error(e)
-        setWalletReceiveAddress(undefined)
-      }
-    })()
-  }, [buyAsset, getReceiveAddressFromBuyAsset])
+          return walletReceiveAddress
+        },
+    staleTime: Infinity,
```

Comment:
blocking: this is *theoretically* very dangerous as a Ledger in diff. configurations/different Ledger devices would have the same deviceId, and `buyAccountId` could well be the same.
Can we use `sellAssetAccountId` in the queryKey as well?

Reply:
Very big 🧠, well caught

Reply:
f5038172fd218eac34922413a5bac20b68046a6d